IFC-1812 Refactor HFID & display_labels

gmazoyer opened the pull request:
This PR brings several changes to how `display_label` and `human_friendly_id` for nodes are stored and retrieved. While it tries to change the least amount of code to keep code compatible as much as possible, it now stores the `display_label` and `human_friendly_id` of nodes into the database to be able to retrieve them when looking up nodes instead of computing them every times. It also introduces events to help in `display_label` computation.

This PR provides implementations for IFC-1820, IFC-1821, IFC-1822, IFC-1823, IFC-1824, IFC-1825 and IFC-1826.

@coderabbitai ignore

gmazoyer committed:
Add `display_label` to base node schema

gmazoyer committed:
Validate display label value

gmazoyer committed:
Add missing utility function

gmazoyer committed:
Add missing utility function

gmazoyer committed:
Add `display_label` for core definitions

gmazoyer committed:
Add `display_label` for core definitions

gmazoyer committed:
Add docs

gmazoyer committed:
Raise error if both display labels are defined

Remove new display label from core schema for now.

gmazoyer committed:
Raise error if both display labels are defined

Remove new display label from core schema for now.

gmazoyer committed:
Always fail validation

ogenstad commented on backend/infrahub/core/schema/schema_branch.py:
```diff
@@ -751,6 +753,26 @@ def validate_uniqueness_constraints(self) -> None:
+            node_schema = self.get(name=name, duplicate=False)
+
+            if node_schema.display_label and node_schema.display_labels:
+                raise ValidationError(f"{node_schema.kind}: cannot defined both `display_label` and `display_labels`")
```
Small typo here "defined" -> "define"

ogenstad commented on backend/infrahub/core/schema/schema_branch.py:
```diff
@@ -751,6 +753,26 @@ def validate_uniqueness_constraints(self) -> None:
+        for name in self.all_names:
+            node_schema = self.get(name=name, duplicate=False)
+
+            if node_schema.display_label and node_schema.display_labels:
```
I think for clarity it should probably be:

```diff
-             if node_schema.display_label and node_schema.display_labels:
+             if node_schema.display_label and node_schema.display_labels is not None:
```

I.e. to ensure that we don't have an empty array.

ogenstad commented on backend/infrahub/core/schema/schema_branch.py:
```diff
@@ -1136,6 +1158,47 @@ def validate_computed_attributes(self) -> None:
+            SchemaElementPathType.ATTR_WITH_PROP
+            | SchemaElementPathType.REL_ONE_MANDATORY_ATTR_WITH_PROP
+            | SchemaElementPathType.REL_ONE_ATTR_WITH_PROP
+        )
```
It could be that we want additional allowed options here later.

gmazoyer committed:
Fix validation

gmazoyer committed:
Fix unit test

gmazoyer committed:
Store the display_label of nodes in the database

gmazoyer committed:
Store the display_label of nodes in the database

gmazoyer committed:
Fix merge with display_label

gmazoyer committed:
Fix merge with display_label

ogenstad commented on backend/infrahub/core/utils.py:
```diff
@@ -214,6 +214,14 @@ def build_regex_attr(value: str | int | bool) -> str:
+
+    This is done by looking at Jinja2 control structures such as `{{ variable }}` or `{% if ... %}`.
+    """
+    return bool(re.search(r"{{.+}}", value) or re.search(r"{%.+%}", value))
```
As this will be used for the display_label where anything outside of Jinja2 would be single fields such as `name__value` I think we can change this slightly to be any string that includes `{` or `}` that way we'd also find invalid templates such as `{ name__value|upper }}` which should be simpler to validate when we try to process this as a Jinja2 template.

gmazoyer committed:
Fix display_label not updated after updating node

gmazoyer committed:
Fix display_label not updated after updating node

gmazoyer committed:
Use branch supprot type from schema

ogenstad committed:
Add events for display labels

gmazoyer committed:
Add `DisplayLabel` class

This class is used to keep track of the display label on a per node
basis. It registers which fields are used to build the display label to
be able to know if the display label needs to be updated when the node
is.

gmazoyer committed:
Add `DisplayLabel` class

This class is used to keep track of the display label on a per node
basis. It registers which fields are used to build the display label to
be able to know if the display label needs to be updated when the node
is.

gmazoyer committed:
Revert a tweak

gmazoyer committed:
Remove unused code in validation

gmazoyer committed:
Consider is jinja2 if curly brace found

gmazoyer committed:
Consider is jinja2 if curly brace found

gmazoyer committed:
Create display label only on node create

gmazoyer committed:
Replicate display_label behaviour for HFID

gmazoyer committed:
Replicate display_label behaviour for HFID

gmazoyer committed:
Fix merge

gmazoyer committed:
Fix merge

gmazoyer committed:
Refactor some code to avoid duplicated code blocks

gmazoyer committed:
Refactor some code to avoid duplicated code blocks

gmazoyer committed:
Rename variables

gmazoyer committed:
Rename variables

gmazoyer committed:
Add methods to update display_label/hfid

gmazoyer committed:
Fix var ref

gmazoyer committed:
Do not fail HFID on enum

gmazoyer committed:
Return stored HFID if it exists

gmazoyer committed:
Add `get_display_label` method like `get_hfid`

gmazoyer committed:
Extract in constants

gmazoyer committed:
Extract in constants

gmazoyer committed:
Ensure classic display label is not broken

gmazoyer committed:
Ensure HFID is set for nodes

gmazoyer committed:
Revert changes to tag schema

gmazoyer committed:
Revert changes to tag schema

gmazoyer committed:
Adapt diff related tests

gmazoyer committed:
Add test with new display label set

gmazoyer committed:
Add test with new display label set

gmazoyer committed:
Fix when using enum in plain display label

gmazoyer committed:
Fix when using enum in plain display label

gmazoyer committed:
Test correct function

gmazoyer committed:
Fix graphql default value for display label if not provided

gmazoyer committed:
Adapted more tests

ajtmccarty commented on backend/tests/unit/core/schema_manager/test_manager_schema.py:
```diff
@@ -1332,6 +1353,47 @@ async def test_validate_display_labels_error(schema_all_in_one, display_labels,
+@pytest.mark.parametrize(
+    "display_label,expected_error",
+    [
+        ("{{ mybool }}", ""),
```
is there no error message for this one?

gmazoyer committed:
More tweaks

gmazoyer committed:
Fix module path

ogenstad committed:
Corrections for tests

gmazoyer committed:
Fix merge of schema nodes

gmazoyer committed:
Fix merge of schema nodes

gmazoyer committed:
Fix expected error messages in tests

gmazoyer committed:
Fix expected error messages in tests

gmazoyer committed:
More fixes for display label validation

gmazoyer committed:
One more adapted test

gmazoyer committed:
Loosen "is jinja2" logic

ogenstad committed:
Add missing parts for display label mutation

gmazoyer committed:
Useless function

gmazoyer committed:
Add is_schema_node property to EnrichedDiffNode

gmazoyer committed:
Add is_schema_node property to EnrichedDiffNode

gmazoyer committed:
Use `set_` instead of `update_`

gmazoyer committed:
Replace ugly `!= "Schema"`

ogenstad committed:
Add mutation and events for HFID

gmazoyer committed:
Use `get_display_label` instead of `render_display_label`

gmazoyer committed:
Ensure path is not used twice in HFID

gmazoyer committed:
Replace `.value` by `.get_value()` and use generic typing

gmazoyer committed:
Replace `.value` by `.get_value()` and use generic typing

gmazoyer committed:
Fix call to `get_display_label`

gmazoyer committed:
await it

gmazoyer committed:
Use `get_hfid()` instead of `.value`

gmazoyer committed:
Integrate display_label and HFID in changelog

gmazoyer committed:
Integrate display_label and HFID in changelog

gmazoyer committed:
Missed unexpected keyword

gmazoyer committed:
Missed unexpected keyword

gmazoyer committed:
Adapt test results

ogenstad committed:
Finalize HFID events

gmazoyer committed:
Put back display_label for node changelog

gmazoyer committed:
Put back display_label for node changelog

ogenstad committed:
Convert old display label format to new when processing the schema

gmazoyer committed:
Remove unneeded async

gmazoyer committed:
Consider update if fields is not provided

gmazoyer committed:
Add migration to trigger HFID/display label backfill (first shot)

ogenstad committed:
Update display label/hfid triggers for local nodes

gmazoyer committed:
Add migration to add `display_label` in schema

Split the backfill migration into its own file.

gmazoyer committed:
Allow updating HFID/display label only

gmazoyer committed:
Fix migration adding hfid/display_label to nodes

Also add a test to validate the migrations 41+42 combo

gmazoyer committed:
Also add display_label to schema generic

gmazoyer committed:
Save updated nodes only once

gmazoyer committed:
Attribute schema are supposed to be set as optional

gmazoyer committed:
Fix mypy errors

gmazoyer committed:
Fix fallback values for HFID/display label

gmazoyer committed:
Fix typing of `get_display_label`

gmazoyer committed:
HFID should be part of changelog indeed

gmazoyer committed:
Allow merging hfid/display label + adapated tests

gmazoyer committed:
Add expected result for last failing test

gmazoyer committed:
Fix last test

gmazoyer committed:
Add display_label and human_friendly_id to valid input names

gmazoyer committed:
Add schema as instance var

gmazoyer committed:
Seems this escape the merge

gmazoyer committed:
Base migration is now 40

gmazoyer committed:
Try to validate upgrade from 1.4

gmazoyer committed:
Migrate all kinds

gmazoyer committed:
mypy ignore for now

gmazoyer committed:
Add progress bar to migration 41

gmazoyer committed:
Backfill using batches of nodes

gmazoyer committed:
Re-use `get_attribute` in node property attribute

ogenstad committed:
Use correct description for HFID & display label triggers

We need to override the description method for these triggers in order
to get the hash values which allows us to properly trigger on updates to
the schema.

ogenstad committed:
Add latest GraphQL schema file

gmazoyer committed:
Fix upgrade from 1.3
